articles/actions.md

```diff
@@ -0,0 +1,338 @@
+
+It is responsible for
+
+* making a request to the action server
+* optionally monitoring the feedback from the action server
+* optionally requesting that the action server cancel the action execution
+* optionally checking the result of the action received from the action server
+
+## Differences between ROS 1 and ROS 2 actions
+
+### First Class Support
+In ROS 1, actions are implemented as a separate library `actionlib` which is built on top of the client libraries.
+This was done to avoid increasing the work required to create a client library in a new language, but actions turned out to be very important to packages like the navigation stack<sup>[1](#separatelib)</sup>.
+In ROS 2 actions will be included in the client library implementations.
+The work of writing a client library in a new language will be reduced by creating a common implmentation in C.
+
+### Services used for Actions
+
+In ROS 1 actions were implemented using a set of topics under a namespace taken from the action name.
+ROS 1 Services were not used because they are inherently synchronous, and actions need to be asynchronous.
+Actions also needed to send status/feedback and be cancelable.
+In ROS 2 services are asynchronous in the common C implementation, so actions will use a combination of services and topics.
+
+### Goal Identifiers
+
+In ROS 1, Action clients are responsible for creating a goal ID when submitting a goal.
+In ROS 2 the action server will be responsible for generating the goal ID and notifying the client.
+
+The server is better equiped to generate a unique goal id than the client because there may be multiple clients who could independently generate the same goal id.
+Another reason is to avoid a race condition between goal creation and cancellation that exists in ROS 1.
+In ROS 1 if a client submits a goal and immediatly tries to cancel it then the cancelation may or may not happen.
+If the cancelation is processed before the goal submission then `actionlib` will ignore the cancellation request without notifying the user's code.
+
+### Namespacing of Generated Messages and Services
+
+Multiple message and service definitions are generated from a single action definition.
+In ROS 1, the generated messages were prefixed with the name of the action to avoid conflicts with other messages and services.
+In ROS 2, the generated service and message definitions will exist in a different namespace to be impossible to conflict with non-action message and service definitions.
+For example, in Python the code from the generated definitions should be in the module `action` instead of `srv` and `msg`.
+In C++, the generated code should be in the namespace and folder `action` instead of `srv` and `msg`.
+
+### Visibility of Action Services and Topics
+
+In ROS 1 `rostopic list` would show all action topics in its output.
+In ROS 2 `ros2 topic list` and `ros2 service list` will not show topics and services used by actions by default.
```

Comment:
will there be a `ros2 action list`??

Reply:
Yes, this is mentioned in the section on [introspection tools](https://github.com/ros2/design/blob/actions_proposal/articles/actions.md#introspection-tools). I can add a pointer to it here.